### PR TITLE
Avoid copy on Refit

### PR DESCRIPTION
**Pull request by cbourjau**
The C-API exposes [`LGBM_BoosterRefit`](https://github.com/microsoft/LightGBM/blob/master/src/c_api.cpp#L2060) which receives the predicted leaf indices as a flat buffer that is laid out in the shape of `nrow x ncol`. The `Boosting::RefitTree` function, on the other hand, expects these indices as a nested vector object (`std::vector<std::vector<int>>`). Creating this nested object requires various additional allocations and amounts to an entire copy of the initial buffer doubling the memory requirements. 

This PR changes the API of `Boosting::RefitTree` to take a pointer to a flat buffer of `int32`s, just like the C-API, thus avoiding the copy. This assumes that this part of the API is not regarded as stable. The C-API remains unchanged.

### CODE REVIEW

**Review by jameslamb — DISMISSED** [review]
Thanks for your interest in LightGBM. It'll be at least a few days until someone's able to review this, as we're currently focused on finishing a release (https://github.com/microsoft/LightGBM/pull/6439#issuecomment-2162120784).

**Comment by jameslamb** [comment]
> `[trigger ci]`

Every CI run on this PR will require a maintainer manually approving it, because you've never contributed here before. Sorry for the inconvenience, but GitHub introduced this as a security measure a few years ago and we've decided to leave it enabled. We do occasionally receive malicious pull requests trying to use our CI resources 🙃

**Comment by cbourjau** [comment]
Thanks for the feedback! Some CI does appear to run, though, and some of it exhibited IO failures the first time around which I was trying to address 🤷 .

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:233]: [nit]

```suggestion
    for (int i = 0; i < nrow); ++i) {
```

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:250]: `pred_leaf.front()` has undefined behavior if the vector is empty :eyes:

```suggestion
    boosting_->RefitTree(pred_leaf.data(), nrow, pred_leaf.size() / nrow);
```

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:235]: I am rather confident that this will break when using `#pragma omp parallel`. IMO, there are two possibilities:

- Get rid of the parallelism (which likely results in a considerable performance hit)
- Initialize `pred_leaf` to equal the final size and use index setters. I'm not sure if this is feasible. One idea would be the following: subsequent code assumes that each line in the text file that is being read provides the same number of values. We could read line 0 which yields the number of values per line. Then, we can resize the vector as needed and we can parallelize reading of the remaining lines. If the value count in a line doesn't match the expected count, we can raise an exception.

File: src/boosting/gbdt.cpp
Comment (borchero) [inline src/boosting/gbdt.cpp:253]: Let's add the previous check as well
```suggestion
  CHECK_GT(nrow * ncol, 0);
  CHECK_EQ(num_data_, nrow);
```

File: include/LightGBM/boosting.h
Comment (borchero) [inline include/LightGBM/boosting.h:77]: Maybe we should "keep" `size_t` for `nrow` and `ncol`?

**Review by borchero — CHANGES_REQUESTED** [review]
Nice, looking forward to this performance improvement! I left a few comments, esp. regarding the changes to the application code

File: src/application/application.cpp
Comment (cbourjau) [reply src/application/application.cpp]: I am pre-allocating the vector now and am using a nested for-loop. Do you still see an issue?

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:242]: Doesn't this need to be

```suggestion
	pred_leaf[irow * ncol + i_row_item] = line_vec[i_row_item];
```

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:233]: I'm unsure what this is doing. Don't we need to split by `\t` to know the number of columns?

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:240]: I think that we should raise an exception here if `line_vec.size() != ncol`

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:257]: This could be simpler, I think

```suggestion
    boosting_->RefitTree(pred_leaf.data(), nrow, ncol);
```

File: src/application/application.cpp
Comment (borchero) [inline src/application/application.cpp:239]: The [compilation error in the CI](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=16409&view=logs&j=e0c3ed1e-0c79-508f-2647-33e12d6ed3d4&t=5d0879b4-8100-51c6-0dab-aba0d8459adc&l=598) teaches me that `irow` must be a signed integer :smile:

**Review by borchero — CHANGES_REQUESTED** [review]
Thanks for the updates, I think we're getting there! 😁

File: src/application/application.cpp
Comment (cbourjau) [reply src/application/application.cpp]: Thanks! I changed the relevant type in gbdt.cpp, too.

**Comment by borchero** [comment]
Thanks @cbourjau! AppVeyor CI will be fixed by https://github.com/microsoft/LightGBM/pull/6490, that only leaves the linting job 😁

File: src/application/application.cpp
Comment (borchero) [reply src/application/application.cpp]: No, looks good now :)

**Review by borchero — APPROVED** [review]
Nice! Thanks a lot for your contribution @cbourjau 🚀

**Review by jameslamb — COMMENTED** [review]
Thanks for this!

I'll dismiss my prior blocking review so it doesn't block merging. But I would like @shiyu1994 or @guolinke to also approve before we merge this. They might remember a reason that this copy was used in #1124.

File: src/application/application.cpp
Comment (shiyu1994) [inline src/application/application.cpp:236]: Shall we use `resize` instead of `reserve`?

**Review by shiyu1994 — APPROVED** [review]
Thanks for the contribution. This should be very helpful.

**Pull request merged**
